docs/src/interface.md

```diff
@@ -1,28 +1,143 @@
-# The algorithm interface
+```@meta
+CollapsedDocStrings = true
+```
+
+# [The algorithm interface](@id sec_interface)
+
+This section starts a single, cohesive story that will weave through all documentation pages.
+We will incrementally build an iterative algorithm, enrich it with stopping criteria, and
+finally refine how it records (logs) its progress. Instead of presenting the API in the
+abstract, we anchor every concept in one concrete, tiny example you can copy & adapt.
+
+Why an “interface” for algorithms? Iterative numerical methods nearly always share the
+same moving pieces:
+
+* immutable input (the mathematical problem you are solving),
+* immutable configuration (parameters and knobs of the chosen algorithm), and
+* mutable working data (current iterate, caches, diagnostics) that evolves as you step.
+
+Bundling these together loosely without forcing one giant monolithic type makes it easier to:
+
+* reason about what is allowed to change and what must remain fixed,
+* write generic tooling (e.g. stopping logic, logging, benchmarking) that applies across many algorithms,
+* test algorithms in isolation by constructing minimal `Problem`/`Algorithm` pairs, and
+* extend behavior (add new stopping criteria, new logging events) without rewriting core loops.
+
+The interface in this package formalizes those roles with three abstract types:
+* [`Problem`](@ref): immutable, algorithm‑agnostic input data.
+* [`Algorithm`](@ref): immutable configuration and parameters deciding how to iterate.
+* [`State`](@ref): mutable data that evolves (current iterate, caches, counters, diagnostics).
+It provides a framework for decomposing iterative methods into small, composable parts:
+concrete `Problem`/`Algorithm`/`State` types have to implement a minimal set of core functionality,
+and this package helps to stitch everything together and provide additional helper functionality such as stopping criteria and logging functionality.
+
+## [Concrete example: Heron's method](@id sec_heron)
+
+To make everything tangible, we will work through a concrete example to illustrate the library's goals and concepts.
+Our running example is Heron's / Babylonian method for estimating $\sqrt{S}$.
+(see also the concise background on Wikipedia: [Babylonian method (Heron's method)](https://en.wikipedia.org/wiki/Methods_of_computing_square_roots#Babylonian_method)):
+Starting from an initial guess $x_0$, we may converge to the solution by iterating:
+
+$$x_{k+1} = \frac{1}{2}\left(x_k + \frac{S}{x_k}\right)$$
```

Comment:
````suggestion
```math
x_{k+1} = \frac{1}{2}\left(x_k + \frac{S}{x_k}\right)
```
````

Is it ok if we do markdown math instead of LaTeX math? 
Not urgent to change that here, just something I would maybe unify when I work through the docs in a later PR.